File: Object Pool/PHP/ThreadPool.php
<?php

require_once "Thread.php";

class ThreadPool {
    private array $threads;
    private int $counter;

    public function __construct()
    {
        $this->threads = [];
        $this->counter = 2;

        for ($i=1; $i <= 2 ; $i++) { 
            $thread = new Thread($i);

            $this->threads[] = $thread;
        }
    }

    public function get() {
        $thread = null;

        if (count($this->threads) > 0) {
            $thread = array_shift($this->threads);
        } else {
            $this->counter++;

            $thread = new Thread($this->counter);
        }

        return $thread;
    }

    public function release(Thread $thread) {
        $this->threads[] = $thread;
    }
}

File: Object Pool/Dart/ThreadPool.dart
import 'Thread.dart';

class ThreadPool {
  List<Thread> _threads = [];
  int _counter = 2;

  ThreadPool() {
    for (var i = 1; i <= 2; i++) {
      _threads.add(Thread(i));
    }
  }

  Thread get() {
    if (_threads.isEmpty) {
      _counter++;
      Thread thread = Thread(_counter);

      return thread;
    } else {
      return _threads.removeAt(0);
    }
  }

  void release(Thread thread) {
    _threads.add(thread);
  }
}
